### PR TITLE
Bunch-Kaufman factorization support for generic number types and inertia computations

### PR DESCRIPTION
**Pull request by djturizo**
### Introduction
This PR adds a generic implementation of the Bunch-Kaufman factorization in native Julia code, and a generic implementation of a inertia calculation function. Right now Julia only support the Bunch-Kaufman factorization for `Float32`, `Float64` and their complex variants. This is because the factorization is handled by LAPACK, which only supports these types. To extend support to generic number types, I translated the LAPACK implementation to native Julia code, and the code performs the factorization in-place. I also included the function `inertia` that computes the number of positive, negative, and zero eigenvalues of an $n \times n$ Bunch-Kaufman factorized matrix in $\mathcal{O}(n)$ time.

### Changes
- `bunchkaufman` and `bunchkaufman!` now work for any `AbstractFloat`, `Rational` and their complex variants. Behavior for previously supported types is not changed (LAPACK is used when possible). `bunchakaufman!` does not support `Integer` types, as in general the factorization lies in the arithmetic closure of the number type (the rationals for the integers). On the other hand, `bunchakaufman` supports `Integer` types, by making an internal conversion to `Rational{BigInt}`.
- `ldiv!` for a `BunchKaufman` factorization has extended support for generic number types with type stability.
- Previously, Julia extracted the diagonal factor of an $n \times n$ `BunchKaufman` object by making a copy of the matrix and then calling a LAPACK function (`dsyconvf`, `csyconvf`, etc., depending on the number type). This function also computes the triangular factor, so it runs in $\mathcal{O}(n^2)$ time. Now Julia uses a native implementation of the LAPACK function with small modifications, so it computes the diagonal factor in $\mathcal{O}(n)$ time, without making a new copy of the matrix.
- Added the function `inertia` that computes the number of positive, negative and zero eigenvalues of the diagonal factor of an $n \times n$ `BunchKaufman` object, in case that the matrix is real symmetric or Hermitian. For complex symmetric matrices, `inertia` only computes the number of zero eigenvalues of the diagonal factor. `inertia` runs in $\mathcal{O}(n)$ time and only uses arithmetic and real absolute value operations. Therefore, `inertia` can be used for matrices of any generic number type, including `Rational`. In particular, for rational matrices the output of `inertia` is exact (unless a positive tolerance is specified).
- Unit tests of the `BunchKaufman` library has been adapted to handle low precision number types (approximate comparisons with tolerance `sqrt(eps(Float64))` do not make sense when the floating point type is `Float16`, for example). The test-set now also runs on the following types: `Float16, Complex{Float16}, BigFloat, Complex{BigFloat}, Complex{Int}, BigInt, Complex{BigInt}, Rational{BigInt}, Complex{Rational{BigInt}}`. Unit tests for the `inertia` function have been added too.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/bunchkaufman.jl [hunk 1]
```diff
@@ -4,6 +4,15 @@
 ## LD for BunchKaufman, UL for CholeskyDense, LU for LUDense and
 ## define size methods for Factorization types using it.
 
+##----------- Type utilities for generic Bunch-Kaufman implementation ------------
+# Generic real type. Any real number type should able to approximate
+# real numbers, and thus be closed under arithmetic operations.
+# Therefore so Int, Complex{Int}, etc. are excluded.
+ClosedReal = T where T <: Union{AbstractFloat, Rational}
```
Comment (oscardssmith) [inline stdlib/LinearAlgebra/src/bunchkaufman.jl:11]: Why not just use `Real` here?

File: stdlib/LinearAlgebra/src/bunchkaufman.jl [hunk 2]
```diff
@@ -390,3 +411,1225 @@ end
...
+lower triangular part (`uplo=='L'`), or the full storage space is used
+(`uplo=='F'`). If `uplo!='F'` then only the corresponding triangular
+part is updated. The values `'U'` or `'L'` can onlt be used when A is
+square (`N==M`).
+
+`syhe` is a character, either `'S'` or `'H'`, indicating whether the
+symmetric adjoint (`syhe=='S'`, and `y'==tranpose(y)`) or the hermitian
+adjoint (`syhe=='H'`, and `y'==adjoint(y)`) must be used.
+"""
+function generic_adr1!(uplo::AbstractChar,
```
Comment (oscardssmith) [inline stdlib/LinearAlgebra/src/bunchkaufman.jl:469]: rather than following a BLAS interface here use types (i.e. no `uplo` or `syhe`)

**Comment by oscardssmith** [comment]
It's great to have this code in Julia! how does the performance compare? The main general code review I have is that with a translation into julia we are no longer bound by the BLAS interface and should be using the types of matrices rather than characters to get all the different variants. Doing so will reduce overhead and should make the code shorter and cleaner.

File: stdlib/LinearAlgebra/src/bunchkaufman.jl (hunk 1 above)
Comment (djturizo) [reply stdlib/LinearAlgebra/src/bunchkaufman.jl:11]: Basically to exclude `Irrational`. As far as I know, irrationals in Julia are not closed under arithmetic operations, so `π+1` does not return a number of type `Irrational`. I don't how to process irrationals in a type stable fashion, so I was forced to exclude them.

File: stdlib/LinearAlgebra/src/bunchkaufman.jl (hunk 2 above)
Comment (djturizo) [reply stdlib/LinearAlgebra/src/bunchkaufman.jl:469]: I guess you something like `if A.uplo == :U ...`, am I right?

Comment (oscardssmith) [reply stdlib/LinearAlgebra/src/bunchkaufman.jl:469]: just `if A isa UpperTriangular`/`if A isa LowerTriangular`. Similarly, `syhe` is just `if A isa Hermitian`/`if A isa Symmetric`

**Comment by djturizo** [comment]
> It's great to have this code in Julia! how does the performance compare? The main general code review I have is that with a translation into julia we are no longer bound by the BLAS interface and should be using the types of matrices rather than characters to get all the different variants. Doing so will reduce overhead and should make the code shorter and cleaner.

The performance is better for small matrices, which I expected because of the overhead of calling LAPACK. For large matrices LAPACK becomes better. A benchmark on my machine using `Float64` matrices gave this:

### Bunch-Kaufman factorization comparison
| Size     | LAPACK Time (mean ± σ) | Julia Time  (mean ± σ) | Ratio of avg. |
|--------|-----------------------------|--------------------------|---------------|
| 1  | 208.613 ns ± 196.848 ns      | 85.333 ns ± 97.350 ns  | 0.409 |
| 10  | 2.870 μs ±  5.123 μs  |  647.215 ns ± 159.824 ns  | 0.226 |
| 100 | 66.201 μs ± 17.581 μs | 89.851 μs ±  18.124 μs | 1.357 |
| 1000 | 37.607 ms ± 663.131 μs | 41.491 ms ±  2.722 ms | 1.103 |
| 10000 | 5.585275 s | 76.761392 s | 13.744 |

For comparison, I ran the same test but doing LAPACK's LU routine v.s. Julia's LU routine. I obtained this:

### LU factorization comparison
| Size     | LAPACK Time (mean ± σ) | Julia Time  (mean ± σ) | Ratio of avg. |
|--------|-----------------------------|--------------------------|---------------|
| 1  | 121.871 ns ± 104.716 ns   | 62.485 ns ± 90.624 ns  | 0.513 |
| 10  | 1.081 μs ± 320.487 ns  |  465.696 ns ± 81.492 ns  | 0.431 |
| 100 | 497.888 μs ± 85.461 μs | 237.850 μs ±  16.197 μs | 0.478 |
| 1000 | 10.609 ms ± 771.916 μs | 257.558 ms ±   4.258 ms | 24.278 |
| 10000 | 6.818 s | 300.694 s | 13.744 | 44.103 | 

In a sense, the penalty of the generic code for large matrices is less than in other already implemented generic codes. I believe this performance is good. Regarding allocations there's barely any (78.17 KiB, over 2 allocations for `n=10000`, exactly the same as the generic LU code allocations).

File: stdlib/LinearAlgebra/src/bunchkaufman.jl (hunk 2 above)
Comment (djturizo) [reply stdlib/LinearAlgebra/src/bunchkaufman.jl:469]: The code performs the factorization in place, so `A` in general will be `Symmetric` or `Hermitian`. The `uplo` check is done to know which triangular part of `A` stores the data. Comparing with `UpperTriangular` or `Lower Triangular` is not feasible in this case.

**Comment by mikmoore** [comment]
What would it take for this to "just work" with more `Number` types. For example, for `QuaternionF64` elements from `Quaternions.jl` (which may require some tweaks since quaternion multiplication is non-commutative). Would duck typing be worth a shot?

Of course, those extensions could happen in a later PR. But it's worth keeping in mind that these methods may be further generalized in the future.

**Comment by djturizo** [comment]
> What would it take for this to "just work" with more `Number` types. For example, for `QuaternionF64` elements from `Quaternions.jl` (which may require some tweaks since quaternion multiplication is non-commutative). Would duck typing be worth a shot?
> 
> Of course, those extensions could happen in a later PR. But it's worth keeping in mind that these methods may be further generalized in the future.

The symmetric version of the algorithm could work, assuming commutativity carefully handled. That's just my guess though, I may be wrong.

File: stdlib/LinearAlgebra/src/bunchkaufman.jl (hunk 2 above)
Comment (djturizo) [reply stdlib/LinearAlgebra/src/bunchkaufman.jl:467]: I have made the changes where in most functions, except for the main factorization function `bunchkaufman_native!` and its helper functions `generic_adr1!` and `bk_rowcol_swap!`. LAPACK works with the raw data directly and so the `BunchKaufman` object stores the raw data as a generic matrix. The main factorization function works with the raw data, and modifying it to work with `Symmetric`, `UpperTriangular`, etc. wrappers only to extract the raw data to store it in the `BumchKaufman` object feels a bit redundant. Of course, I modified the rest of the functions to work directly with the `BunchKaufman` object if available, so the code has simplified a bit. Is this acceptable?

**Comment by dkarrasch** [comment]
This is awesome! Regarding the row and column permutation, we might as well (re)use some generic code like https://github.com/JuliaLang/julia/pull/50471#issuecomment-1763339586. Perhaps I should make an independent PR so that both PRs can use it.

**Comment by djturizo** [comment]
> This is awesome! Regarding the row and column permutation, we might as well (re)use some generic code like [#50471 (comment)](https://github.com/JuliaLang/julia/pull/50471#issuecomment-1763339586). Perhaps I should make an independent PR so that both PRs can use it.
> 
> Edit: your version here seems to be doing something more sophisticated...

If I understood correctly, the code you linked operates over the whole matrix. The permutation in this PR is done by only modifying the appropriate triangular part of a symmetric/Hermitian matrix, thereby reducing the cost by half (again, this is adapter from LAPACK).

**Comment by dkarrasch** [comment]
Could you please fix the typos? Then we should be ready to go. Perhaps this should be announced in the NEWS.md as a new feature.

**Comment by djturizo** [comment]
> Could you please fix the typos? Then we should be ready to go. Perhaps this should be announced in the NEWS.md as a new feature.

Done. Please let me know of any other fixes needed.

**Comment by dkarrasch** [comment]
In the root folder of the repo, there's a NEWS.md file. Could you please write a quick announcement there in the LinearAlgebra section, together with a reference to this PR?

**Comment by djturizo** [comment]
> In the root folder of the repo, there's a NEWS.md file. Could you please write a quick announcement there in the LinearAlgebra section, together with a reference to this PR?

Done. Is there anything else that I can help with?

File: NEWS.md [hunk 3]
```diff
@@ -115,6 +115,7 @@ Standard library changes
    respectively, now efficiently compute the generalized eigenvalues (`eigen`: and eigenvectors) of `A` and `B`. Note: The second
    argument is the output of `bunchkaufman` or `lu` ([#50471]).
 * Structured matrices now retain either the axes of the parent (for `Symmetric`/`Hermitian`/`AbstractTriangular`/`UpperHessenberg`), or that of the principal diagonal (for banded matrices) ([#52480]).
+* `bunchkaufman` and `bunchkaufman!` now work for any `AbstractFloat`, `Rational` and their complex variants. `bunchakaufman` now supports `Integer` types, by making an internal conversion to `Rational{BigInt}`. Added new function `inertia` that computes the inertia of the diagonal factor given by the `BunchKaufman` factorization object of a real symmetric or Hermitian matrix. For complex symmetric matrices, `inertia` only computes the number of zero eigenvalues of the diagonal factor ([#51487]).
```
Comment (dkarrasch) [inline NEWS.md:118]:
```suggestion
* `bunchkaufman` and `bunchkaufman!` now work for any `AbstractFloat`, `Rational` and their complex variants. `bunchkaufman` now supports `Integer` types, by making an internal conversion to `Rational{BigInt}`. Added new function `inertia` that computes the inertia of the diagonal factor given by the `BunchKaufman` factorization object of a real symmetric or Hermitian matrix. For complex symmetric matrices, `inertia` only computes the number of zero eigenvalues of the diagonal factor ([#51487]).
```

**Comment by dkarrasch** [comment]
Thank you very much! What a great contribution. Welcome to the Julia community @djturizo.

**Pull request merged**